src/crypto/hsm/nxp/PersistentStorageOperationalKeystoreHSM.cpp

```diff
@@ -0,0 +1,279 @@
+    keyidFabIdMapping[slotId].keyId       = mPendingKeypair->GetKeyId();
+    keyidFabIdMapping[slotId].fabricIndex = mPendingFabricIndex;
+    keyidFabIdMapping[slotId].isPending   = 0;
+
+    // If we got here, we succeeded and can reset the pending key: next `SignWithOpKeypair` will use the stored key.
+    ResetPendingKey();
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR PersistentStorageOperationalKeystoreHSM::RemoveOpKeypairForFabric(FabricIndex fabricIndex)
+{
+    uint8_t i = 0;
+    VerifyOrReturnError(mStorage != nullptr, CHIP_ERROR_INCORRECT_STATE);
+    VerifyOrReturnError(IsValidFabricIndex(fabricIndex), CHIP_ERROR_INVALID_FABRIC_INDEX);
+
+    ChipLogProgress(Crypto, "SE05x: RemoveOpKeypair for Fabric %02x", fabricIndex);
+
+    // Remove pending state if matching
+    if ((mPendingKeypair != nullptr) && (fabricIndex == mPendingFabricIndex))
+    {
+        RevertPendingKeypair();
+    }
+
+    while (i++ < MAX_KEYID_SLOTS_FOR_FABRICS)
+    {
+        if (keyidFabIdMapping[i].fabricIndex == fabricIndex)
+        {
+            // Delete the keyPair associated with the fabric
+            keyidFabIdMapping[i].isPending = 0;
+            Platform::Delete<Crypto::P256KeypairHSM>(keyidFabIdMapping[i].pkeyPair);
+            keyidFabIdMapping[i].pkeyPair    = NULL;
+            keyidFabIdMapping[i].keyId       = 0;
+            keyidFabIdMapping[i].fabricIndex = 0;
+        }
+    }
+
+    return CHIP_NO_ERROR;
+}
+
+void PersistentStorageOperationalKeystoreHSM::RevertPendingKeypair()
+{
+    ChipLogProgress(Crypto, "SE05x: RevertPendingKeypair");
+    VerifyOrReturn(mStorage != nullptr);
+    // Just reset the pending key, we never stored anything
+    ResetPendingKey();
+}
+
+CHIP_ERROR PersistentStorageOperationalKeystoreHSM::SignWithOpKeypair(FabricIndex fabricIndex, const ByteSpan & message,
+                                                                      Crypto::P256ECDSASignature & outSignature) const
+{
+    uint8_t i = 0;
+    ChipLogProgress(Crypto, "SE05x: SignWithOpKeypair");
+
+    if (mIsPendingKeypairActive && (fabricIndex == mPendingFabricIndex))
+    {
+        VerifyOrReturnError(mPendingKeypair != nullptr, CHIP_ERROR_INTERNAL);
+        // We have an override key: sign with it!
+        ChipLogProgress(Crypto, "SE05x: SignWithOpKeypair ==> using mPendingKeypair");
+        return mPendingKeypair->ECDSA_sign_msg(message.data(), message.size(), outSignature);
+    }
+    else
+    {
+        while (i++ < MAX_KEYID_SLOTS_FOR_FABRICS)
+        {
+            if ((keyidFabIdMapping[i].fabricIndex == fabricIndex) && (keyidFabIdMapping[i].isPending == 0))
+            {
+                ChipLogProgress(Crypto, "SE05x: SignWithOpKeypair ==> using stored keyPair");
+                return keyidFabIdMapping[i].pkeyPair->ECDSA_sign_msg(message.data(), message.size(), outSignature);
+            }
+        }
+    }
+
+    ChipLogProgress(Crypto, "SE05x: SignWithOpKeypair ==> No keyPair found");
+    return CHIP_ERROR_INTERNAL;
```

Comment:
Shouldn't this be CHIP_ERROR_INVALID_FABRIC_INDEX?

Reply:
Changed.